File: 450-delete-node-in-a-bst/delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        def delete(node, target):
            if not node:
                return None
            if target > node.val:
                node.right = delete(node.right, target)
            elif target < node.val:
                node.left = delete(node.left, target)
            else:
                if not node.left:
                    return node.right
                if not node.right:
                    return node.left
                else:
                    cur = node.left
                    while cur.right:
                        cur = cur.right
                    node.val = cur.val
                    node.left = delete(node.left, node.val)
            return node
        
        return delete(root, key)
            

File: 450-delete-node-in-a-bst/delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) {
            return root;
        }
        if (root.val == key) {
            if(root.right == null) {
                return root.left;
            }
            if(root.left == null) {
                return root.right;
            }
            TreeNode cur = root.right;
            while (cur.left != null) {
                cur = cur.left;
            }
            root.val = cur.val;
            root.right = this.deleteNode(root.right, root.val);
        }
        else if (root.val < key) {
            root.right = this.deleteNode(root.right, key);
        } 
        else {
            root.left = this.deleteNode(root.left, key);
        }
        return root;
    }
}